File: 0113_path_sum_II.py
# n = len(root)
# Time: O(nlogn)
# Space: O(nlogn)
from typing import Optional, List


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def pathSum(
            self, root: Optional[TreeNode], targetSum: int
    ) -> List[List[int]]:
        if not root:
            return []
        ans = []
        stack = [(root, targetSum - root.val, [root.val])]

        while stack:
            curr, val, ls = stack.pop()
            if not curr.left and not curr.right and val == 0:
                ans.append(ls)
            if curr.left:
                stack.append(
                    (curr.left, val - curr.left.val, ls + [curr.left.val])
                )
            if curr.right:
                stack.append(
                    (curr.right, val - curr.right.val, ls + [curr.right.val])
                )
        return ans

File: go/0113_path_sum_II.go
package main

import "slices"

// n = number of nodes in the tree
// Time:  O(n)
// Space: O(n)
func pathSum(root *TreeNode, targetSum int) [][]int {
	results := [][]int{}

	var dfs func(*TreeNode, []int, int)

	dfs = func(node *TreeNode, curPath []int, curTarget int) {
		if node == nil {
			return
		}

		curPath = append(curPath, node.Val)
		curTarget -= node.Val
		if node.Left == nil && node.Right == nil {
			if curTarget == 0 {
				results = append(results, slices.Clone(curPath))
			}
		}

		dfs(node.Left, curPath, curTarget)
		dfs(node.Right, curPath, curTarget)
	}

	dfs(root, []int{}, targetSum)

	return results
}
